primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -0,0 +1,635 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+//! Houses the code that implements the transactional overlay storage.
+
+use super::{StorageKey, StorageValue};
+
+use itertools::Itertools;
+use std::collections::{HashSet, BTreeMap, BTreeSet};
+use smallvec::SmallVec;
+
+const PROOF_DIRTY_KEYS: &str = "\
+	We assume transactions are balanced. Every start of a transaction created one dirty
+	keys element. This function is only called on transaction close. Therefore an element
+	created when starting the transaction must exist; qed";
+
+const PROOF_DIRTY_OVERLAY_VALUE: &str = "\
+	A write to an OverlayedValue is recorded in the dirty key set. Before an OverlayedValues
+	is removed its containing dirty set is removed. This function is only called for keys that
+	are in the dirty set. Therefore the entry must exist; qed";
+
+const PROOF_OVERLAY_NON_EMPTY: &str = "\
+	An OverlayValue is always created with at least one transaction and dropped as soon
+	as the last transaction is removed; qed";
+
+type DirtyKeys = SmallVec<[HashSet<StorageKey>; 5]>;
+type Transactions = SmallVec<[InnerValue; 5]>;
+
+#[derive(Debug, Default, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+struct InnerValue {
+	/// Current value. None if value has been deleted.
+	value: Option<StorageValue>,
+	/// The set of extrinsic indices where the values has been changed.
+	/// Is filled only if runtime has announced changes trie support.
+	extrinsics: BTreeSet<u32>,
+}
+
+/// An overlay that contains all versions of a value for a specific key.
+#[derive(Debug, Default, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct OverlayedValue {
+	// The individual versions of that value.
+	// One entry per transactions during that the value was actually written.
+	transactions: Transactions,
+}
+
+/// Holds a set of changes with the ability modify them using nested transactions.
+#[derive(Debug, Default, Clone)]
+pub struct OverlayedChangeSet {
+	/// Stores the changes that this overlay constitutes.
+	changes: BTreeMap<StorageKey, OverlayedValue>,
+	/// Stores which keys are dirty per transaction. Needed in order to determine which
+	/// values to merge into the parent transaction on commit. The length of this vector
+	/// therefore determines how many nested transactions are currently open (depth).
+	dirty_keys: DirtyKeys,
```

Comment:
I opted for the simple solution that might be a bit less optimized on purpose because the previous PR did bot get merged because of its complexity. As long as its good enough I think we should keep it that way and make further optimizations in a follow up PR. Otherwise this would end up in another 3k comments PR.

Reply:
I have no idea what is 'good enough', I am just feeling a bit bad about the double tree set.
This little optimization does not sound like too much complexity, but yeah I can be a bad judge (on one hand you need to understand is_dirty_key function, on the other you need to compare two index, but yes the alternative solution gives room to a user wrongly looking for a key in the dirty key set).

Reply:
Actually this optimization is not as good as it seems, I oversee the merging of dirty_keys on commit  (will need to iterate on the big btreeset for dirty key state, this could even be slower in some limit case (very big set and small number of change in the layer so you hop more than with merging two close to empty btreeset, but since we accumulate change from every layer the bottom layer will be merging some big set these case should get amortized). Not that straightforward, and is only really good at getting good micro benchmarks (does not change the fact that number of transaction impacts write cost).

Reply:
I am mainly concerned with correctness here to have understandable code that can actually pass a review. Be my guest to integrate clearly justified optimizations in a follow up PR. For that reason I am closing all remarks that suggest complicating the code for performance reasons.